file upload endpoints

majestrate opened the pull request:
add room file upload and room file fetching endpoints with unit tests

Fixes #65 #80

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -264,3 +269,39 @@ def message_unpin(room, msg_id):
+    filename = request.headers.get("X-Filename")
+    if not filename:
+        abort(http.BAD_REQUEST)
+    id = room.upload_file(request.data, g.user, filename=secure_filename(filename))
```
Don't need `secure_filename` here: `room.upload_file` already takes care of sanitizing it.

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -264,3 +269,39 @@ def message_unpin(room, msg_id):
+    if not room_file:
+        abort(http.NOT_FOUND)
+
+    resp = make_response(send_from_directory(abspath(curdir), room_file.path, as_attachment=True))
```
Switch to `send_file` instead so that you can specify the path and `download_filename` separately, because basename(path) != filename: the stored filename has a prefix of `{id}_`, and will be shortened if the given filename was long.  (We don't need send_from_directory's path protections because `path` is sanitized before we insert it into the database).

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -264,3 +269,39 @@ def message_unpin(room, msg_id):
+
+    resp.headers["Date"] = http_date(room_file.uploaded)
+    resp.headers["Expires"] = http_date(room_file.expiry)
+    resp.headers["Content-Type"] = 'application/octet-stream'
```
Should we attempt to guess the content-type when we have a filename with a known extension?  Presumably there is some Python module that makes that easy.

jagerman commented on tests/request.py:
```diff
@@ -10,16 +10,24 @@ def sogs_get(client, url, user):
     return client.get(url, headers=x_sogs_for(user, "GET", url))
 
 
+def sogs_post_raw(client, url, data, user, *, ctype='application/octet-stream', extra_headers=None):
```
```suggestion
def sogs_post_raw(client, url, data, user, *, ctype='application/octet-stream', extra_headers={}):
```

jagerman commented on tests/request.py:
```diff
@@ -10,16 +10,24 @@ def sogs_get(client, url, user):
+    for k in extra_headers or []:
+        if k not in headers:
+            headers[k] = extra_headers[k]
+    return client.post(url, data=data, content_type=ctype, headers=headers)
```
```suggestion
    return client.post(url, data=data, content_type=ctype, headers={**headers, **extra_headers})
```
Simplification, and also I don't think the `not in headers` makes sense here: if I give an extra header value that conflicts with one of the X-SOGS headers then I'd expect it to override when the header already exists rather than silently being ignored.

jagerman commented on tests/test_room_routes.py:
```diff
@@ -6,7 +6,8 @@
 from util import pad64
-from request import sogs_get, sogs_post, sogs_put
+from request import sogs_get, sogs_post, sogs_put, sogs_post_raw
+import secrets
```
```suggestion
from nacl.utils import random
```
Not a big deal, just a minor preference for libsodium's `randombuf_bytes`, partly because `token_bytes` is a stupid name.

jagerman commented on tests/test_room_routes.py:
```diff
@@ -1041,3 +1042,42 @@ def test_edits(client, room, user, user2, mod, global_admin):
+
+    filename = 'testfile.bin'
+    url_post = f"/room/{room.token}/file"
+    file_content = secrets.token_bytes(1024)
```
```suggestion
    file_content = random(1024)
```

jagerman commented on tests/test_room_routes.py:
```diff
@@ -1041,3 +1042,42 @@ def test_edits(client, room, user, user2, mod, global_admin):
+    r = sogs_post_raw(
+        client,
+        url_post,
+        secrets.token_bytes(1024),
```
```suggestion
        random(1024),
```

jagerman commented on tests/test_room_routes.py:
```diff
@@ -1041,3 +1042,42 @@ def test_edits(client, room, user, user2, mod, global_admin):
+
+    r = sogs_get(client, f'/room/test-room/file/{id}/{filename}', user)
+    assert r.status_code == 200
+    assert r.data == file_content
```
Some other things to add to the tests:

- test that the file exists on disk in the right place (`uploads/test-room/1_{filename}`)
- for a simple filename check that we are getting `Content-Disposition` set to `attachment; filename="testfile.bin"`.  (Or possibly flask is producing `attachment; filename*=UTF-8''testfile.bin`; if so, use that for the test instead).
- upload with unicode characters in the filename (currently we strip that out of the actual path/filename we write to disk (which is good), but I think we might also be stripping it out of the `filename`, which we shouldn't).  E.g. `my 🎂.txt` should end up as (I think) `uploads/test-room/2_my_.txt` for path, `my 🎂.txt` for filename, and Content-Disposition on the download should be: `attachment; filename*=UTF-8''my%20%F0%9F%8E%82.txt`.
- if a user has `upload=false` set they shouldn't be able to upload an attachment.
- if a user has `write=false` permission (even with `upload` true) they shouldn't be able to upload an attachment.
- a user with `read=false` access should not be able to download an attachment.
- make an upload, pin it to a room (which will set the upload expiry to 0), and then verify that downloading it doesn't include an Expires header.

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -264,3 +269,39 @@ def message_unpin(room, msg_id):
+    if not room.check_upload(g.user):
+        abort(http.FORBIDDEN)
+
+    filename = request.headers.get("X-Filename")
```
This header isn't going to work for unicode filenames; the only standardized encoding of unicode through HTTP headers (AFAIK) is [RFC 5987](https://www.rfc-editor.org/rfc/rfc5987), which requires something like:

```
Foo: v*=UTF-8''xxxxxx
```
where xxxxx is the url-encoded utf-8 byte sequence, for arbitrary `Foo` and `v`.  E.g. in a Content-Disposition header for `my 🎂.txt`:
```
Content-Disposition: attachment; filename*=UTF-8''my%20%F0%9F%8E%82
```

We could do something similar for the `X-Filename` header, but it seems like it'd be just as easy to use the `Content-Disposition` header as above for the upload request rather than `X-Filename`.

(And hopefully werkzeug exposes some means of parsing such unicode-tagged headers).

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -264,3 +269,39 @@ def message_unpin(room, msg_id):
+    resp = make_response(send_from_directory(abspath(curdir), room_file.path, as_attachment=True))
+
+    resp.headers["Date"] = http_date(room_file.uploaded)
+    resp.headers["Expires"] = http_date(room_file.expiry)
```
What happens here if `expiry` is None?

majestrate replied:
will that ever happen?

jagerman replied:
Yup: room images have a null expiry, as will attachments in pinned posts (though that isn't implemented yet).

majestrate replied:
what should we send in this case, expiry is 0?

jagerman replied:
I think just omit the Expiry header.

jagerman replied:
From what I gather, there is no "never" value for Expiry headers, and a browser is supposed to treat anything that it fails to parse as 0 which it is supposed to mean "it expired a long time ago"

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -264,3 +269,39 @@ def message_unpin(room, msg_id):
+
+
+@rooms.get("/room/<Room:room>/file/<int:fileId>/<filename>")
+@auth.user_required
```
```suggestion
@auth.read_required
```

jagerman reviewed (changes requested):
Left various comments above.

One more piece is needed here, though, to avoid orphan uploads by associating uploads with posts in sogs: see #65.

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -301,7 +301,7 @@ def serve_file(room, fileId, filename):
     resp.headers["Date"] = http_date(room_file.uploaded)
-    resp.headers["Expires"] = http_date(room_file.expiry)
-    resp.headers["Content-Type"] = 'application/octet-stream'
+    if room_file.expiry and room_file.expiry > 0:
```
0 is a special value that means "expire now", which isn't something we should really worry about here.  I think, however, that `model.room.Room.get_file` should be taking care of not returning expired files that haven't hit the cleanup job yet, i.e. by adding a not expired check here:
```Python
if row:
    return File(row)
```
i.e.:
```Python
if row and row['expiry'] > time.time():
    return File(row)
```

jagerman replied:
(It'd be even better if it was in the query, except then we'd potentially trigger the extra file_id_hacks query crap unnecessarily).

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -296,11 +296,13 @@ def serve_file(room, fileId, filename):
     if not room_file:
         abort(http.NOT_FOUND)
+    if room_file.expiry == 0:
+        abort(http.NOT_FOUND)
```
See my other comments: I think `model.Room` should be responsible for not returning an expired File rather than routes.

jagerman commented:
One other thing: we should clean up the `upload/test-room` directory as part of the tests DB cleanup (would probably work at the end of the `db` fixture in `conftest.py`).

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -337,3 +339,53 @@ def poll_room_info(room, info_updated):
+@rooms.post("/room/<Room:room>/file")
+@auth.user_required
+def upload_file(room):
+    """ upload a file to a room """
```
Needs to be rebased on dev and the docs moved out of api.yaml to here (updated if needed).

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -337,3 +339,53 @@ def poll_room_info(room, info_updated):
+def serve_file(room, fileId, filename):
+    """
+    serves a file uploaded to a room
+    """
```
Move docs from api.yaml

jagerman commented:
#77 should be ready for merge or cherry-picking now, which adds the `files` arguments to add_post/edit_post to handle updating file ownership.

Commit:
add file upload endpoints

fixes #65

Commit:
allow setting image via PUT /room/<token>

fixes #80

jagerman committed:
Don't require filename

Makes the filename not required (we just put in a null if not provided)
for upload or download.

jagerman committed:
Remove debug

jagerman committed:
Build the file response manually

send_file() has optimizations that won't work through an onion request
(in particular, we can't extract the response body from such a
response, but we need that for onion requests).

This changes it to build a flask.Response that streams it via Python
file iteration instead, which is a bit less efficient, but allows
extracting the body where needed for onion request handling.  (It should
still stream, rather than slam everything into memory, when invoked
directly).

jagerman committed:
Only warn for >= 400 responses

We return some valid non-200 responses (like 201, 304) that shouldn't be
producing log spam.

jagerman committed:
Preserve unicode filenames & fix tests

Changes the file upload to only strip bad characters from the filename
we use to generate the path on disk, but not the suggested filename we
store in the database and return to the user.

Also changes the test suite to interpret the random byte filename as
latin1 so that it ends up as utf-8 for flask.  Otherwise flask does the
latin1 conversion itself, and returns utf-8 back, but that breaks the
test (because what we get back has gone through an unexpected latin1 ->
utf-8 encoding).

jagerman committed:
Make sure we disassociate a room image from an owning post

In practice it is probably already null, but in theory you could upload
a post attachment and then pin that.

jagerman committed:
Fix retrieval of infinite expiry files (like room images)

Commit:
revert sql query to not reference non existing column

operator precedence in comparison fixup

jagerman committed:
File upload test fixes

1. Flask's test client is doing something strange when setting a header
   property starting with \\.  Just bypass it entirely at set the header
   ourself using filename*=UTF-8''... encoding (flask understands this,
   but doesn't produce it).

2. Make random filename tests deterministic (to avoid intermittent
   failures), and increase the count to 500 and length to 32 (instead of
   128 and 16) so that we're (hopefully) hitting more weird cases.

3. Add a specific test with leading backslashes in the filename because
   that specifically was causing issues with flask (see point 1).

4. Replace \0 and / with unicode replacement characters in the uploaded
   filename because those aren't valid filename characters *anywhere*.

5. Do the expected local path filename cleanup and verify that we stored
   at the expected path (even for utf8 or unsafe filenames).

jagerman committed:
Tests: rename seconds/minutes/etc. to from_now.seconds, etc.

It wasn't obvious how seconds()/minutes()/etc. worked (I got it wrong
the first time I tried it), so rename it inside a `from_now` class that
makes it more self-descriptive.

jagerman committed:
Verify that room image upload updates expiry timestamp

jagerman committed:
Older python compat fix

jagerman committed:
Fix broken `.room_id` attribute

If called after `.room` had been accessed it would return None instead
of the id.

jagerman committed:
Add File.post_id attribute

jagerman committed:
Finish off test_owned_files test case implementations

jagerman committed:
Move upload/download API doc from api.yaml

jagerman committed:
Fix & test cross-posting of uploads

Add tests for cross-room file references.  We were correctly handling
not allowing cross-room references of post attachments, but cross-room
references of the room image was allowed (now fixed).
